### PR TITLE
EIP712 Struct hash implementation

### LINKED ISSUE #1448: [Ethereum] Implemented EIP712
Currently it's implemented in both iOS and Android app (v3), need to add v4 support too

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Compute the hash of a struct, used for signing, according to EIP712.  Fixes #1448 .

## Testing instructions

Unit tests, *EthereumAbi*.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Review by Dj-The-closer — APPROVED** [review]

File: swift/Tests/Blockchains/EthereumAbiTests.swift [hunk 1]
```diff
@@ -100,4 +100,51 @@ class EthereumAbiTests: XCTestCase {
...
+                {
+                    "Mail": [
+                        {"name": "from", "type": "Person"},
+                        {"name": "to", "type": "Person[]"},
+                        {"name": "contents", "type": "string"}
+                    ]
+                }
+            ]
+        """
+        let hash = EthereumAbiStruct.hashStruct(structType: structType, valueJson: value, typesJson: types)
```
Comment (hewigovens) [inline swift/Tests/Blockchains/EthereumAbiTests.swift:147]: Here is one json rpc request sample:

```json
{
  "method": "eth_signTypedData_v4",
  "params": [
    "0x9d8a62f656a8d1615c1294fd71e9cfb3e4855a4f",
    "{\"domain\":{\"chainId\":1,\"name\":\"Ether Mail\",\"verifyingContract\":\"0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC\",\"version\":\"1\"},\"message\":{\"contents\":\"Hello, Bob!\",\"from\":{\"name\":\"Cow\",\"wallets\":[\"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826\",\"0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF\"]},\"to\":[{\"name\":\"Bob\",\"wallets\":[\"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB\",\"0xB0BdaBea57B0BDABeA57b0bdABEA57b0BDabEa57\",\"0xB0B0b0b0b0b0B000000000000000000000000000\"]}]},\"primaryType\":\"Mail\",\"types\":{\"EIP712Domain\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"version\",\"type\":\"string\"},{\"name\":\"chainId\",\"type\":\"uint256\"},{\"name\":\"verifyingContract\",\"type\":\"address\"}],\"Group\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"members\",\"type\":\"Person[]\"}],\"Mail\":[{\"name\":\"from\",\"type\":\"Person\"},{\"name\":\"to\",\"type\":\"Person[]\"},{\"name\":\"contents\",\"type\":\"string\"}],\"Person\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"wallets\",\"type\":\"address[]\"}]}}"
  ],
  "id": 1621843180588
}
```

second param is stringified json
```json
{
  "domain": {
    "chainId": 1,
    "name": "Ether Mail",
    "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
    "version": "1"
  },
  "message": {
    "contents": "Hello, Bob!",
    "from": {
      "name": "Cow",
      "wallets": [
        "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826",
        "0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF"
      ]
    },
    "to": [
      {
        "name": "Bob",
        "wallets": [
          "0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB",
          "0xB0BdaBea57B0BDABeA57b0bdABEA57b0BDabEa57",
          "0xB0B0b0b0b0b0B000000000000000000000000000"
        ]
      }
    ]
  },
  "primaryType": "Mail",
  "types": {
    "EIP712Domain": [
      { "name": "name", "type": "string" },
      { "name": "version", "type": "string" },
      { "name": "chainId", "type": "uint256" },
      { "name": "verifyingContract", "type": "address" }
    ],
    "Group": [
      { "name": "name", "type": "string" },
      { "name": "members", "type": "Person[]" }
    ],
    "Mail": [
      { "name": "from", "type": "Person" },
      { "name": "to", "type": "Person[]" },
      { "name": "contents", "type": "string" }
    ],
    "Person": [
      { "name": "name", "type": "string" },
      { "name": "wallets", "type": "address[]" }
    ]
  }
}
```
As you can see, types and values info are inside one json, I'd like to have interface like this: `let hash/data =EthereumAbi.encodeTyped(json: jsonString)`

Comment (catenocrypt) [reply swift/Tests/Blockchains/EthereumAbiTests.swift]: API updated

**Comment by catenocrypt** [comment]
API renamed (EthereumAbi.encodeTyped).

**Comment by catenocrypt** [comment]
Coverage increased.

**Comment by catenocrypt** [comment]
Ready for review.

**Review by hewigovens — APPROVED** [review]
Nice work, only left 2 comments

File: include/TrustWalletCore/TWEthereumAbi.h [hunk 2]
```diff
@@ -31,4 +31,42 @@ bool TWEthereumAbiDecodeOutput(struct TWEthereumAbiFunction* _Nonnull fn, TWData
...
+///     },
+///     "message": {
+///         "name": "Cow",
+///         "wallet": "CD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"
+///     }
+///  })");
+/// On error, empty Data is returned.
+/// Returned data must be deleted (hint: use WRAPD() macro).
+TW_EXPORT_STATIC_METHOD
+TWData* _Nonnull TWEthereumAbiEncodeTyped(TWString* _Nonnull messageJson);
```
[the remark this answers is not in the document]
Comment (vikmeup) [reply include/TrustWalletCore/TWEthereumAbi.h:70]: we would support only v4 right?

[the remark this answers is not in the document]
Comment (hewigovens) [reply include/TrustWalletCore/TWEthereumAbi.h:70]: we should support both v3/v4, `eth_signTypedData` and `eth_signTypedData_v4` -> v4, `eth_signTypedData_v3` -> v3

**Review by Dj-The-closer — APPROVED** [review]

File: include/TrustWalletCore/TWEthereumAbi.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (catenocrypt) [reply include/TrustWalletCore/TWEthereumAbi.h:70]: v3 does not support arrays, recursive types, v4 is a superset of v3.  Therefore I think we don't need to support v3.

**Comment by catenocrypt** [comment]
I have realized one step is missing for signature (putting together hash of EIP712Domain and message).  I'm switch this back to Draft, will add missing part.

File: include/TrustWalletCore/TWEthereumAbi.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (hewigovens) [reply include/TrustWalletCore/TWEthereumAbi.h:70]: As long as it produces same hash for v3 rpc call, I'm ok with this

**Comment by catenocrypt** [comment]
Ready for review (2nd time)

File: tests/Ethereum/AbiStructTests.cpp [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+    EXPECT_EQ(hex(msgMailCow1Bob1.hashStruct()), "c52c0ee5d84264471806290a3f2c4cecfc5490626bf912d01f240d7a274b371e");
+
+    EXPECT_EQ(hex(msgEIP712Domain.hashStruct()), "f2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f");
+
+    Address address = Address(privateKeyCow.getPublicKey(TWPublicKeyTypeSECP256k1Extended));
+    EXPECT_EQ(address.string(), "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826");
+
+    // TODO sig
+    // 0xbe609aee343fb3c4b28e1df9e632fca64fcfaede20f02e86244efddf30957bd2
+    // 0x4355c47d63924e8a72e509b65029052eb6c299d53a04e167c5775fd466751c9d07299936d304c153f6443dfa05f40ff007d72911b6f72307f996231605b915621c
```
[the remark this answers is not in the document]
Comment (catenocrypt) [reply tests/Ethereum/AbiStructTests.cpp]: Verification of signature added.

**Review by hewigovens — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]

**Pull request merged**